keps/sig-api-machinery/1693-warnings/README.md

```diff
@@ -0,0 +1,209 @@
+-->
+
+- [x] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] Design details are appropriately documented
+- [x] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] Graduation criteria is in place
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This enhancement makes it easier for users and cluster administrators
+to recognize and remedy use of deprecated APIs.
+Users are presented with informative warnings at time of use.
+Administrators are given metrics that show deprecated API use,
+and audit annotations that can be used to identify particular API clients.
+
+## Motivation
+
+Kubernetes has many deprecations in flight at any given moment, in various stages, with various time horizons.
+Keeping track of all of them is difficult, and has historically required careful reading of release notes,
+and manually sweeping for use of deprecated features.
+
+### Goals
+
+* When a user makes a request to a deprecated API, present them with a warning
+  that includes the target removal release and any replacement API
+* Allow a cluster administrator to programatically determine if deprecated APIs are being used:
+  * Filtered to particular APIs
+  * Filtered to particular operations (e.g. get, list, create, update)
+  * Filtered to APIs targeting removal in particular releases
+* Allow a cluster administrator to programatically identify particular clients using deprecated APIs
+
+### Non-Goals
+
+While the proposed warning mechanism is generic enough to carry arbitrary warnings,
+the following items are out of scope for the first iteration of this feature:
+
+* Allowing extensions mechanisms like admission webhooks to contribute warnings
+* Surfacing warnings about other non-fatal problems
+  (for example, [problematic API requests](http://issue.k8s.io/64841#issuecomment-395141013)
+  that cannot be rejected for compatibility reasons)
+
+## Proposal
+
+### Server-side changes
+
+When a deprecated API is used:
+
+1. Add a `Warning` header to the response
+2. Increment a counter metric with labels for the API group, version, resource, API verb, and target removal major/minor version
```

Comment:
Where do we store target removal versions? Isn't that also something we need to build?

Reply:
also part of https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/1635-prevent-permabeta